Improve Service controller's code coverage a little bit

dhilipkumars opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Improves the code coverage for Service Controller
Before
```
go test --cover ./pkg/controller/service
ok      k8s.io/kubernetes/pkg/controller/service        0.101s  coverage: 23.4% of statements
```
After
```
go test --cover ./pkg/controller/service/
ok      k8s.io/kubernetes/pkg/controller/service        0.094s  coverage: 62.0% of statements
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
More unit testing

**Special notes for your reviewer**:

**Release note**: 
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```

Commit:
Improve Service controller code coverage a little bit

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41763)
<!-- Reviewable:end -->

cmluciano commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+			},
+			expectedErr: fmt.Errorf("unexpected key format: %q", "invalid/key/string"),
+		},
+		/* We cannot open this test case as syncService(key) currently runtime.HandleError(err) and suppresses frequently occurring errors
```
Should this be removed?

dhilipkumars replied:
will remove it, and replace it with a single sentence TODO

cmluciano commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+			printError = true
+			if obtainedErr != nil && tst.expectedErr != nil {
+				if obtainedErr.Error() == tst.expectedErr.Error() {
+					//Both are valid error objects but still their content is same, then not an error
```
This sentence does not read well to me. Can you rephrase to clarify your intent? The comment might also be unnecessary.

cmluciano commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		{
+			testName: "If an invalid service is deleted",
+			updateFn: func() {
+				//srv = defaultExternalService()
```
Need to remove?

dhilipkumars replied:
agreed

cmluciano commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+				expectedError := "Service external-balancer not in cache even though the watcher thought it was. Ignoring the deletion."
+				if srvErr == nil || srvErr.Error() != expectedError {
+					//cannot be nil or Wrong error message
```
Comment may be uneccesary

dhilipkumars replied:
agreed

cmluciano commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
 }
 
 // TODO(a-robinson): Add tests for update/sync/delete.
```
Can probably remove this TODO if this is merged

dhilipkumars replied:
agreed

cmluciano reviewed (commented):
Mostly nits:
- I think a lot of the code is readable enough that the comments may not be necessary [Ex. struct comments](https://github.com/kubernetes/kubernetes/pull/41763/files#diff-d57aa6e65ab02bdc723f13236834a5d8R322).
- Some portions are commented out as either not working or are values that may have been used for testing. These should probably be removed in place of TODOs if someone should take a stab at adding them later.

dhilipkumars commented:
@cmluciano and @bprashanth 
Shall i also remove this function? Its not being used anywhere
```
func intSlicesEqual(x, y []int) bool {
	if len(x) != len(y) {
		return false
	}
	if !sort.IntsAreSorted(x) {
		sort.Ints(x)
	}
	if !sort.IntsAreSorted(y) {
		sort.Ints(y)
	}
	for i := range x {
		if x[i] != y[i] {
			return false
		}
	}
	return true
}
```

dhilipkumars commented:
CC other owners for more feedback:@MrHohn @bowei @thockin @mrubin

dhilipkumars commented:
PING, Hello authors @bprashanth @bowei @thockin @MrHohn @mrubin any feedback on this PR? Its been a month

dhilipkumars commented:
@k8s-bot ok to test

MrHohn commented:
@k8s-bot ok to test

/assign

MrHohn commented:
@dhilipkumars Terribly sorry about the delay. I will review your PR soon.

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+				controller, cloud, _ = newController()
+
+			},
+			expectedErr: fmt.Errorf("unexpected key format: %q", "invalid/key/string"),
```
This might be risky, it will be broken whenever client-go package changes the error format. Probably enough to just check that it returns error?

dhilipkumars replied:
agreed removed, i was thinking if we have more than once situation where it throws error how do we differentiate it, so far we have only one case that throws error

MrHohn replied:
Fair point, though I'm still concern about the consistence of error message, which depends on other package. Or could we make this a boolean for now and put a TODO?

dhilipkumars replied:
Done, Added TODO and chacked it to only check for error.  This helps a little more for us to print what was expected and what was obtained.

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+	}
+
+	for _, tst := range testCases {
+		var printError bool
```
This looks weird to me, could we figure out a simpler way to log error?

dhilipkumars replied:
Simplified now

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
+	}{
+		{
+			testName: "If an invalid service is deleted",
```
If a non-exist service is deleted

dhilipkumars replied:
done

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+	testCases := []struct {
+		testName        string
+		updateFn        func()                                                //Update function used to manupulate srv and controller values
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
```
You are using expectedFn elsewhere, why not make it consistent here?
```
expectedFn func(error, time.Duration) error
```

dhilipkumars replied:
done

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+			testName: "If delete was successful",
+			updateFn: func() {
+
+				controller, cloud, _ = newController()
```
Seems like newController() is needed for all test cases. Why not move it out of the function and pass it in asan argument?

dhilipkumars replied:
done

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName        string
```
Why not also have a key field here, just like TestUpdateService?

dhilipkumars replied:
I have moved the key as a global variable as we are using the same key throughout the test cases.

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+	for _, tst := range testCases {
+		tst.updateFn()
+		obtainedErr, retryDuration := controller.processServiceDeletion("external-balancer")
```
Add the key field and pass it in here.

dhilipkumars replied:
moved the key as global variable so that it can be changed

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+	testCases := []struct {
+		testName       string //Name of the test case
+		updateFn       func() //Function to update the old and new service varuables
```
s/varuables/variables

Or "Function to update the service object"

MrHohn replied:
The way you are using updateFn here is not intuitive. Usually I would expect oldService and newService as fields in the testcase struct. Though I could see making this a function make it simpler for tweaking both old and new services. We could keep it for now, but it will be great if you have better idea :)

dhilipkumars replied:
done

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+	testCases := []struct {
+		testName       string //Name of the test case
+		updateFn       func() //Function to update the old and new service varuables
+		expectedResult bool   //needsupdate always returns bool
```
expectedResult -> expectNeedsUpdate, or something that is more obvious

dhilipkumars replied:
done

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+	}{
+		{
+			testName: "If the service type is different",
```
I think this test name is not accurate? Changes like ClusterIP -> NodePort will not trigger loadBalancer update. Probably just say change service type from LoadBalancer to ClusterIP?

dhilipkumars replied:
done

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+func TestServiceCache(t *testing.T) {
+
+	sC := &serviceCache{serviceMap: make(map[string]*cachedService)} //ServiceCache
```
nit: sC -> sc

MrHohn replied:
All the testcases will be sharing this one cache, is this intended? And doe the order of test cases matter? In that case, I think we should at least put a comment so folks won't break it.

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+			},
+		},
+		{
+			testName: "SetandGet",
```
nit: SetandGet -> Set and Get

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		nArray[i].Name = fmt.Sprintf("node1")
+	}
+	if !nodeSlicesEqualForLB(nArray, nArray) {
+		t.Errorf("Expected=true Obtained=false")
```
Could we also put the function name (nodeSlicesEqualForLB) in the error log to make it clearer?

dhilipkumars replied:
done, but thought it would come anyways when you run the test with -v?

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+	testCases := []struct {
+		testName   string
+		runTest    func()
+		expectedFn func() error
```
According to how you use runTest and expectedFn, I think they should have more appropriate names. Probably setCacheFn and checkCacheFn (I'm also bad at naming :()?

dhilipkumars replied:
_more appropriate names. Probably setCacheFn and checkCacheFn__
Agreed.
 _(I'm also bad at naming :()_
Disagreed.  You are far better. :-)

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		},
+		{
+			testName: "allServices",
+			runTest: func() {
```
Probably set it to nil, and check for nil and do nothing in the test case loop?

dhilipkumars replied:
great idea, done

MrHohn reviewed (changes requested):
Thanks for the PR and improving the our coverage rate :)

dhilipkumars commented:
Thanks for the review i will update the PR addressing your comments.  Thanks.  I will also try include [this](https://github.com/kubernetes/kubernetes/commit/5c9fe166f98136b15a32a7f281d995d1e8a44b63) commit changes,  this PR gives an additional of 15% coverage.

dhilipkumars commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -61,6 +63,20 @@ func newController() (*ServiceController, *fakecloud.FakeCloud, *fake.Clientset)
 	return controller, cloud, client
 }
 
+func defaultExternalService() *v1.Service {
```
done

dhilipkumars commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
 }
 
-// TODO(a-robinson): Add tests for update/sync/delete.
+func TestUpdateService(t *testing.T) {
```
done

dhilipkumars commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		testName   string
+		key        string
+		updateFn   func() //Manupulate the structure
+		srv        *v1.Service
```
done throughout the tests

dhilipkumars commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+		},
+	}
+
+	for _, tst := range testCases {
```
done throughout the tests

dhilipkumars commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+	}
+}
+
+func TestDeleteService(t *testing.T) {
```
done

dhilipkumars commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+
+}
+
+// Add unit testing for needsUpdate functions
```
done

dhilipkumars commented:
@MrHohn 
Now after merging [this](https://github.com/kubernetes/kubernetes/commit/5c9fe166f98136b15a32a7f281d995d1e8a44b63) commit we get 3% improved coverage. 
```
time go test --cover ./pkg/controller/service/
ok      k8s.io/kubernetes/pkg/controller/service        0.092s  coverage: 65.0% of statements

real    1m41.993s
user    2m53.844s
sys     0m18.272s

```
PTAL now.

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -650,95 +628,100 @@ func TestDoesExternalLoadBalancerNeedsUpdate(t *testing.T) {
 
+//All the testcases for ServiceCache uses a single cache, these below test cases should be run in order,
+//as tc1 (addCache would add elements to the cache)
+//and tc2 (delCache woudl remove element from the cache without it adding automatically)
```
s/woudl/would

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -314,139 +307,112 @@ func TestGetNodeConditionPredicate(t *testing.T) {
+			},
+		},
+		{
+			testName: "If Updating a  new Loadbalancer IP",
```
tiny nit: "If Updating a  new Loadbalancer IP" -> "If updating LoadBalancer IP"

dhilipkumars replied:
Agreed

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -484,54 +450,58 @@ func TestSyncService_new(t *testing.T) {
+		obtainedErr := controller.syncService(tc.key)
+
+		//Lets only check if we expect error but recived nil and viceversa.
+		if (obtainedErr != nil && tc.expectedErr == nil) && (obtainedErr == nil && tc.expectedErr != nil) {
```
Should use ||?
```
if (obtainedErr != nil && tc.expectedErr == nil) || (obtainedErr == nil && tc.expectedErr != nil)
```

dhilipkumars replied:
Oops, that is correct.

MrHohn commented on pkg/controller/service/servicecontroller_test.go:
```diff
@@ -484,54 +450,58 @@ func TestSyncService_new(t *testing.T) {
 }
 
-func TestDeleteService(t *testing.T) {
+func TestProcessDeleteService(t *testing.T) {
```
Make it TestProcessServiceDeletion to be consistent with the origin function name.

MrHohn reviewed (commented):
Thanks! Mostly looks good, put some nits and comments.

MrHohn commented:
BTW, could you change the release note to:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```

So the robot will replace the release note label.

dhilipkumars commented:
@k8s-bot 
Release Notes: None

dhilipkumars commented:
```
BTW, could you change the release note to:
```
Should i be part of the Org to be able to do that? Im unable to change or add any lables.

MrHohn commented:
> Should i be part of the Org to be able to do that? Im unable to change or add any lables.

No, you don't need to in this case. Once you change the release note in your first post to NONE, bot will apply label for you. As I can see now the release note is still empty, instead of NONE.

MrHohn reviewed (approved):
Thanks! LGTM. Could you squash the commits?

dhilipkumars commented:
done now

MrHohn commented:
/lgtm

MrHohn commented:
/assign @thockin for approval

MrHohn commented:
/assign @thockin
for approval.

thockin commented:
/approve
